sbpy/photometry/core.py

```diff
@@ -507,12 +549,104 @@ class DiskIntegratedModelClass(Fittable1DModel):
     Bond albedo is 0.0184
     >>> print('Phase integral is {0:.3}'.format(phaseint))
     Phase integral is 0.368
+
+    Initialization with subclass of `~sbpy.data.DataClass`:
```

Comment:
This. Is. So. Cool! Thanks!

Reply:
But then we have a problem here:  Now one model class object could contain multiple phase function models, it will calculate the magnitudes for multiple targets at once.  The question is, how do save the magnitudes of multiple targets into one single `Ephem` object for output?

Reply:
Following the astropy docs, the `mag` method should allow multiple `Ephem` objects:
> When evaluating a model set, by default the input must be the same length as the number of models, with one input per model.

If only one is provided, then return multiple `Ephem` objects.

Are there use cases where returning a modified input object will be awkward?  Let the model set represent V, R, and I phase functions.  It wouldn't make sense to return three different ephemeris objects.  Let the `Ephem` object contain multiple asteroids.  If the user instantiated the model with it, then wouldn't they expect the `mag` method to apply the appropriate model to the corresponding lines in `eph`?  Also, what if the `Ephem` object already has a magnitude column?

I think a simpler approach is to return `Quantity` object(s) that the user can append as a column on their own.

Reply:
@mkelley Right, allowing model sets would make things extremely complicated, which we should avoid I believe in order to ensure the robustness of our module.  I do think we want to allow model sets because it would eventually facilitate fitting multiple asteroids/bands all at once.  This is in fact the whole point of model sets:
> Model sets are used primarily for fitting, allowing a large number of models of the same type to be fitted simultaneously (and independently from each other) to some large set of inputs.

So I tend to agree with your suggestion to go back to just returning `Quantity` object(s) and let user decide how to use them.  What do you think, @mommermi ?

Reply:
I follow the reasoning of @mkelley - up to the final part :) I encountered the same problem in `sbpy.thermal` and took the following approach:
* for now, a model class instantiation can only contain a single object; `sbpy.thermal` hence takes a single-line `Phys` object and one `Ephem` object corresponding to the former
* the fitting functions (`.fit` and `.from_data` like in `sbpy.photometry`) return an updated `Phys` object; the `.flux` method returns an updated `Ephem` object

"Updated" means here that information resulting from the methods are added to the input object tables. I prefer this approach as it preserves the order of the data and minimizes confusion. Especially in the case of thermal (and probably also photometric) data, the resulting properties (like flux) are often also only meaningful if the other input parameters are available at the same time. Yes, people could append a `Quantity` object easily to a table, but I think that is unnecessary overhead. Whatever you want to do with the resulting properties (e.g., plotting), you will always need the other properties that are present in the input object. Hence, I think we should always keep everything together and simply append the results. 

Does this make sense?

Reply:
What is your approach for a spectrum?  If I input 500 wavelengths, then the ephemeris will come back with 500 extra columns?

Reply:
I am not adding a lot of columns here. Instead - in the case of `sbpy.thermal` - I simply add two columns (`thermal_lam` for wavelengths and `thermal_flux` for fluxes) and then project the corresponding data on these columns. I wrote a `DataClass.expand` function just for this purpose (https://github.com/mommermi/sbpy/blob/thermal/sbpy/data/core.py#L521). 

In the case of a single spectrum with 500 wavelengths, the table will have a length of 500 and a minimum of 2 columns (if we drop all ephemeris information, which could be stored separately).

Reply:
Ah, so you are dropping the ephemeris data.  Why return an `Ephem` object?

Reply:
I'm not dropping them in the case of `thermal` data, since there's usually not a lot of data points - I guess the same applies to `photometry` data. I'm just saying that we could drop the ephemeris data in the case of a spectrum. Simply because the overhead would be big.

Reply:
There's not a lot of data points, unless you are like me and work with a lot of data points :)  For example, we have ground-based mid-IR spectra of comet 46P, and we want to have the nucleus removed before we model the dust composition.  Or, consider IRTF/SpeX LXD spectroscopy of an NEA.  The thermal emission will be significant throughout the entire wavelength range.  I recommend a keyword argument that controls whether or not a Quantity is returned.  My preference would be to have Quantity the default, because that is the simple case.

I do like the ability to have flux densities and such appended to ephemeris objects.  In fact, I have already been doing something similar in mskpy.  There, all my models returned flat arrays, but higher-level objects could generate lightcurve tables from ephemeris data.  Having that option in sbpy is a must.

This has me thinking about combining models.  We will want to combine thermal and reflected light, accounting for phase effects and reddening, and then use the combined model to fit photometry, a spectrum, or lightcurve.  Given all this subclassing and overloading, it will still work, right?

Reply:
Thanks guys for the suggestions.  Looks like we have to defer to the future for a full resolution of the issues involved here.  This can be a good topic of discussion when we all get together in June.  For v0.1 release, how about this approach:

1. We instantiate a model class using only the first valid parameter set found in the input `DataClass` object.  We do not deal with multiple objects and model set at this time.
1. We return an `Ephem` object with the high level methods, such as `.mag`, `.ref` in photometry model classes.  At least I don't have to deal with the multiple wavelength cases with one model class like in @mommermi 's `thermal` module :)

Please let me know if this sounds good to you for v0.1 @mkelley @mommermi

Reply:
Ok, so how about this:
* as @mkelley suggested, functions and methods can return `Quantity` objects if they only calculate a single property. If they calculate several properties, they should be bundled in a `DataClass` table. Even if a function/method only returns a single property, there should always be an option to add the newly derived properties to the corresponding input `DataClass` object (if any)
* as @jianyangli  suggested, let's only support single object models for now.

Can we all agree on this approach? If so, I can update the wiki page.

Concerning combining models: this will be tricky. `sbpy.thermal` is currently supporting phase effects only for the NEATM. I have been thinking a lot about how to setup a model for reflected solar light and albedo calculations using phase integrals. For now, I will all this hard code into my code for testing, but the goal is to use existing or build new models for these cases. That is something we can also aim for in June.

Reply:
Sounds good to me.  I'll update the code to follow this approach.  Do you have a suggestion for the keyword name to return a `DataClass` object so we can all use the same name across the whole `sbpy` package?

Reply:
how about `append_results`?

Reply:
Yes, let's take this route for now and discuss more in June.